### DISCUSSION TITLE
Permit 'unsafe' modifier differences between partial declarations

### DISCUSSION THREAD

**Discussion by RikkiGibson** [dotnet/csharplang#6650] [Language Ideas]
@jkoritzinsky has pointed out that partial implementations sometimes want to use unsafe code, and using an unsafe block inside the body is unpleasant.

```cs
partial void M();

partial void M()
{
    unsafe
    {
        int* ptr = GetUnsafeThing();
        UseUnsafeThing(ptr); // etc...
    }
}
```

Source generator authors would rather simply be able to include the unsafe modifier only on the implementation part. It's also felt that generated code should stylistically match what users would actually write as much as possible, and users often prefer the unsafe modifier over unsafe blocks.

```cs
partial void M();

unsafe partial void M() // currently an error, but perhaps should be relaxed
{
    int* ptr = GetUnsafeThing();
    UseUnsafeThing(ptr); // etc...
}
```

**Comment by HaloFour** [comment]
Is it really that necessary to optimize the experience for generated code?  Writing the source generator to emit the `unsafe` block is trivial.

**Reply by jkoritzinsky** [reply]
Here's the issue with using an `unsafe` block: It's very non-idiomatic C#, and the authors for the source generators in the BCL, including myself, have a strong goal of making our generated code as idiomatic as possible. In particular, for the interop source generators, we have a goal that developers can and would feel comfortable using the generated code as a starting point and editing from there, and the .NET team has gotten significant feedback that using `unsafe` blocks for the whole method body is "ugly" and that very few teams use that or want to use it.

The interop generator as of now just puts the `unsafe` modifier on the containing type of the method for simplicity, since there is no requirement to keep `unsafe` consistent between `partial` type declarations.

**Reply by jnm2** [reply]
I'm someone that likes to use `unsafe` scoped tightly, even if it means surrounding the whole method body, if the method signature itself is not unsafe. "Idiomatic" may not be universal on this topic.

**Comment by jkoritzinsky** [comment]
Here's some notes from our conversation that led to this discussion:

Pros:

- Using an `unsafe` block for a whole method body instead of putting the `unsafe` keyword in the signature is "ugly" and not the way C# is commonly written.
- The presence of the `unsafe` modifier does not need to be consistent between partial type declarations, only partial method declarations.

Cons:

- Generator authors can just put `unsafe` on the containing type, which is what the interop generators do today.
- There's been some consideration of making `unsafe` more viral and giving it additional meaning than a syntactic switch (like tying it to https://github.com/dotnet/runtime/issues/31354). If we were to go down this route, then `unsafe` would have to be consistent since it wouldn't be just a syntactic switch. However, this change would be a large breaking change, so I don't know if LDM has an appetite for something like it anyway.

**Comment by CyrusNajmabadi** [comment]
A compelling reason to allow this is that the *signature* of the partial method may itself contain unsafe types (like `int*`).  Generators can solve this by putting `unsafe` at the type level.  But conceptually this makes the most sense at the method declaration level.

**Reply by jnm2** [reply]
That would not constitute an unsafe modifier *difference*  between partial parts, because the signature containing the pointer type is present on all parts and all parts would need to be in an unsafe lexical scope.

**Reply by CyrusNajmabadi** [reply]
Agreed :-)
